### PR TITLE
Autonomous

### PR DESCRIPTION
**Pull request by codeNinjaDev**
Autonomous No Competition Routines.

### CODE REVIEW

**Comment by codeNinjaDev** [comment]
@chackotaco Please review

File: src/Auto/Action/DriveIntervalAction.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include "Timer.h"
+#include "DriveIntervalAction.h"
+
+
+
+DriveIntervalAction::DriveIntervalAction(DriveController* kDrive, double seconds, double y, double x) {
+  goal_time = seconds;
+  x_drive = x;
+  y_drive = y;
+  this->kDrive = kDrive;
```
Comment (ajorgens) [inline src/Auto/Action/DriveIntervalAction.cpp]: Why the name `kDrive`? There may be a better variable name for this.

File: src/Auto/Action/ShootAction.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  bool IsFinished();
+  void Update();
+  void Done();
+  void Start();
+  virtual ~ShootAction();
+protected:
+
+
+private:
+  double kSpeed;
```
Comment (ajorgens) [inline src/Auto/Action/ShootAction.h]: Again, why k in front of the variable name?

File: src/Auto/Auto.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+void Auto::ListOptions() {
+  selector->ListOptions();
+}
+
+void Auto::Start() {
+  //Get the dashboard selection and set the auto routine
+  autoRoutine = selector->Pick();
+  autoRoutineRunner->setAutoRoutine(autoRoutine);
+
```
Comment (ajorgens) [inline src/Auto/Auto.cpp]: Extra whitespace

File: src/Auto/Auto.cpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ *  Created on: Jan 28, 2017
+ *      Author: peter
+ */
+#include "Timer.h"
+#include "Auto.h"
+#include <iostream>
+#include <stdio.h>
+using namespace std;
+
+Auto::Auto(DriveController* kDrive, RobotModel* kShooter) {
```
Comment (ajorgens) [inline src/Auto/Auto.cpp:14]: (next PR) We want to add a function to `Auto` to be called from `MainProgram`. If `Teleop.init()` is called, we want to call a function that will call `Done` for all actions.

File: src/Auto/Auto.cpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  autoRoutine = selector->Pick();
+  autoRoutineRunner->setAutoRoutine(autoRoutine);
+
+
+
+  //run the prestart for the function
+  autoRoutine->Prestart();
+  //Run the set autoroutine
+  autoRoutineRunner->Start();
+  //stop(); //!! Check if this is necessary!!
```
Comment (ajorgens) [inline src/Auto/Auto.cpp]: Is this necessary?

File: src/Auto/AutoRoutine.cpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+}*/
+
+void AutoRoutine::Run() {
+  m_active = true;
+  /*try {
+    routine();
+  }*/
+
+  Routine();
+
```
Comment (ajorgens) [inline src/Auto/AutoRoutine.cpp]: Lots of extra whitespace

File: src/Auto/AutoRoutine.cpp [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+#include "AutoRoutine.h"
+/*void AutoRoutine::WaitTime(double seconds) {
+  AutoRoutine::runAction(new TimeoutAction(seconds));
+}*/
+
+void AutoRoutine::Run() {
+  m_active = true;
+  /*try {
```
Comment (ajorgens) [inline src/Auto/AutoRoutine.cpp]: Remove commented out code

File: src/Auto/AutoRoutine.cpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  m_active = false;
+}
+
+bool AutoRoutine::IsActive() {
+  return m_active;
+}
+
+void AutoRoutine::RunAction(Action* action) {
+  action->Start();
+  while (IsActive() && !action->IsFinished()) {
```
Comment (ajorgens) [inline src/Auto/AutoRoutine.cpp]: `IsActive()` -> `m_active`

File: src/Auto/AutoRoutine.cpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+
+}
+
+void AutoRoutine::Stop() {
+  m_active = false;
+}
+
+bool AutoRoutine::IsActive() {
+  return m_active;
```
Comment (ajorgens) [inline src/Auto/AutoRoutine.cpp:25]: Is this function called anywhere else except this class?

File: src/Auto/AutoRoutine.h [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+class AutoRoutine {
+public:
+  bool m_active = false;
+  virtual void Prestart() = 0;
+  void Run();
+  void Stop();
+  bool IsActive();
+  static constexpr double m_update_rate = (1.0/50.0);
+
+  //bool isActiveWithThrow() throw AutoRoutineEndedException;
```
Comment (ajorgens) [inline src/Auto/AutoRoutine.h]: Needed? If not, remove

File: src/Auto/AutoRoutine.h [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  void RunAction(Action* action);
+  void DriveInterval(DriveController* kDrive, double seconds, double y, double x);
+  void Shoot(RobotModel* kShooter, double seconds, double speed);
+
+protected:
+  virtual void Routine(void) = 0;
+
+  //virtual void routine();
```
Comment (ajorgens) [inline src/Auto/AutoRoutine.h]: remove

File: src/Auto/AutoRoutineSetter.cpp [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+}
+
+void AutoRoutineSetter::Start() {
+
+  if(m_auto_routine != nullptr) {
+
+    m_auto_routine->Run();
+  }
+  autoTimer->Start();
```
Comment (ajorgens) [inline src/Auto/AutoRoutineSetter.cpp]: Why call start if `m_auto_routine` is a `nullptr`? Should this be inside the if call?

File: src/Auto/AutoRoutineSetter.cpp [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+void AutoRoutineSetter::Stop() {
+  if(m_auto_routine != nullptr) {
+    m_auto_routine->Stop();
+  }
+  autoTimer->Stop();
```
Comment (ajorgens) [inline src/Auto/AutoRoutineSetter.cpp]: Same comment.

File: src/Auto/AutoRoutineSetter.h [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+#ifndef SRC_AUTO_AUTOROUTINESETTER_H_
+#define SRC_AUTO_AUTOROUTINESETTER_H_
+
+#include "AutoRoutine.h"
+
+class AutoRoutineSetter {
+public:
+
+  //void function NOT constructor
```
Comment (ajorgens) [inline src/Auto/AutoRoutineSetter.h]: Remove

File: src/Auto/Routines/DoNothingRoutine.cpp [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ *  Created on: Jan 26, 2017
+ *      Author: peter
+ */
+
+#include "DoNothingRoutine.h"
+
+void DoNothingRoutine::Routine() {
+  SmartDashboard::PutString("Running DoNothing", "DoNothingRoutine");
+	// TODO Auto-generated constructor stubSmartDashboard::PutString("Running DriveRoutine", "DriveRoutine");
```
Comment (ajorgens) [inline src/Auto/Routines/DoNothingRoutine.cpp]: Do we want this logged to the dashboard?

File: src/Auto/Routines/DriveForwardRoutine.cpp [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+
+#include "DriveForwardRoutine.h"
+
+DriveForwardRoutine::DriveForwardRoutine(DriveController* kDrive) {
+
+  this->kDrive = kDrive;
+}
+
+void DriveForwardRoutine::Routine() {
+	// TODO Auto-generated constructor stub
```
Comment (ajorgens) [inline src/Auto/Routines/DriveForwardRoutine.cpp]: remove comment

File: src/Auto/Routines/JustShootRoutine.cpp [hunk 17]
```diff
@@ -0,0 +1,16 @@
+
+
+#include "JustShootRoutine.h"
+
+JustShootRoutine::JustShootRoutine(RobotModel* shooter) {
+  kShooter = shooter;
+}
+
+void JustShootRoutine::Prestart() {
+  // TODO Auto-generated constructor stub
```
Comment (ajorgens) [inline src/Auto/Routines/JustShootRoutine.cpp:9]: Does anything need to be done in `PreStart`?

File: src/ControlBoard.cpp [hunk 18]
```diff
@@ -44,24 +48,24 @@ void ControlBoard::ReadControls() {
...
   if (USING_WIN_DRIVER_STATION) {
     driverLeftJoyX = driverJoy->GetRawAxis(XINPUT_WIN_LEFT_X_AXIS);
-    driverLeftJoyY = driverJoy->GetRawAxis(XINPUT_WIN_LEFT_Y_AXIS);
+    driverLeftJoyY = -driverJoy->GetRawAxis(XINPUT_WIN_LEFT_Y_AXIS);
     driverRightJoyX = driverJoy->GetRawAxis(XINPUT_WIN_RIGHT_X_AXIS);
-    driverRightJoyY = driverJoy->GetRawAxis(XINPUT_WIN_RIGHT_Y_AXIS);
+    driverRightJoyY = -driverJoy->GetRawAxis(XINPUT_WIN_RIGHT_Y_AXIS);
 
-    operatorLeftJoyX = operatorJoy->GetRawAxis(XINPUT_WIN_LEFT_X_AXIS);
+    /*operatorLeftJoyX = operatorJoy->GetRawAxis(XINPUT_WIN_LEFT_X_AXIS);
```
Comment (ajorgens) [inline src/ControlBoard.cpp]: Why is the operator code commented out?

File: src/DashboardLogger.cpp [hunk 19]
```diff
@@ -43,6 +47,36 @@ void DashboardLogger::PutDriveMotorCurrentData() {
   SmartDashboard::PutNumber("PDP_rightDriveMotorA", robot->GetCurrent(RIGHT_DRIVE_MOTOR_A_PDP_CHAN));
   SmartDashboard::PutNumber("PDP_rightDriveMotorB", robot->GetCurrent(RIGHT_DRIVE_MOTOR_B_PDP_CHAN));
 }
+
+void DashboardLogger::PutShooterEncoderData() {
+  SmartDashboard::PutNumber("SHOOTER_ENC_GetDistance()", robot->shooterEncoder->GetDistance());
+  //SmartDashboard::PutNumber("SHOOTER_ENC_GetRate()", robot->shooterEncoder->GetRate());
+  SmartDashboard::PutNumber("SHOOTER_ENC_GetRPM", robot->shooterEncoder->GetRate()*60.0);
+  SmartDashboard::PutNumber("SHOOTER_ENC_GetRPMGRAPH", robot->shooterEncoder->GetRate()*60.0);
+  //SmartDashboard::PutNumber("SHOOTER_ENC_GetPeriod()", robot->shooterEncoder->GetPeriod());
```
Comment (ajorgens) [inline src/DashboardLogger.cpp]: Why are all of these commented out?

File: src/MainProgram.cpp [hunk 20]
```diff
@@ -5,21 +5,26 @@
...
 #include "ControlBoard.h"
 #include "DashboardLogger.h"
+#include "ShooterController.h"
 #include <string.h>
+#include "Auto/Auto.h"
 
 class MainProgram: public frc::IterativeRobot {
   //LiveWindow helps in Test mode
-  LiveWindow *lw;
+  //LiveWindow *lw;
```
Comment (ajorgens) [inline src/MainProgram.cpp:14]: Why is this commented out?

File: src/MainProgram.cpp [hunk 21]
```diff
@@ -63,15 +75,18 @@ class MainProgram: public frc::IterativeRobot {
     currTimeSec = robot->GetTime();
     deltaTimeSec = currTimeSec - lastTimeSec;
 
-    robot->UpdateCurrent();
+    //robot->UpdateCurrent();
```
Comment (ajorgens) [inline src/MainProgram.cpp]: Why is this commented out?

File: src/MainProgram.cpp [hunk 22]
```diff
@@ -81,32 +96,34 @@ class MainProgram: public frc::IterativeRobot {
...
+
     dashboardLogger->UpdateData();
 
     //Updates timer
     lastTimeSec = currTimeSec;
     currTimeSec = robot->GetTime();
     deltaTimeSec = currTimeSec - lastTimeSec;
 
-    robot->UpdateCurrent();
+    //robot->UpdateCurrent();
```
Comment (ajorgens) [inline src/MainProgram.cpp]: And this?

File: src/MainProgram.cpp [hunk 23]
```diff
@@ -81,32 +96,34 @@ class MainProgram: public frc::IterativeRobot {
...
+
     superstructureController->Reset();
 
   }
 
   void DisabledPeriodic() {
     dashboardLogger->UpdateData();
-
-    robot->UpdateCurrent();
+    //robot->UpdateCurrent();
```
Comment (ajorgens) [inline src/MainProgram.cpp]: And this?

File: src/RobotModel.cpp [hunk 24]
```diff
@@ -7,36 +7,59 @@
...
+  rightDriveEncoder->SetReverseDirection(true);
+  rightDriveEncoder->SetDistancePerPulse( ((1.0)/(250.0)) * ((4.0)*(M_PI)) );
+  rightDriveEncoder->SetSamplesToAverage(90);
 
   leftDriveMotorA->SetSafetyEnabled(false);
   leftDriveMotorB->SetSafetyEnabled(false);
   rightDriveMotorA->SetSafetyEnabled(false);
   rightDriveMotorB->SetSafetyEnabled(false);
-  shooterMotor->SetSafetyEnabled(false);
+  /*shooterMotorA->SetSafetyEnabled(false);
```
Comment (ajorgens) [inline src/RobotModel.cpp]: And this?

File: src/ShooterController.cpp [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+
+void ShooterController::Update(double currTimeSec, double deltaTimeSec) {
+  switch (m_stateVal) {
+    case (kInitialize):
+      nextState = kTeleop;
+      break;
+    case (kTeleop):
+      //Shooter Behaviour
+      if(humanControl->GetShooterRunDesired()){
+        if(SHOOTER_USE_PID){
```
Comment (ajorgens) [inline src/ShooterController.cpp]: What is `SHOOTER_USE_PID`?

**Review by ajorgens — CHANGES_REQUESTED** [review]
Initial review completed. Please address comments made

File: src/Auto/Auto.cpp (hunk 5 above)
Comment (codeNinjaDev) [reply src/Auto/Auto.cpp]: Yes it starts the routine and timer.

File: src/Auto/AutoRoutine.cpp (hunk 9 above)
Comment (codeNinjaDev) [reply src/Auto/AutoRoutine.cpp:25]: No

File: src/Auto/Routines/JustShootRoutine.cpp (hunk 17 above)
Comment (codeNinjaDev) [reply src/Auto/Routines/JustShootRoutine.cpp:9]: no

Comment (codeNinjaDev) [reply src/Auto/Routines/JustShootRoutine.cpp:9]: Some routines need it though, I have not made it.

File: src/ControlBoard.cpp (hunk 18 above)
Comment (codeNinjaDev) [reply src/ControlBoard.cpp]: Ask Michael?

File: src/DashboardLogger.cpp (hunk 19 above)
Comment (codeNinjaDev) [reply src/DashboardLogger.cpp]: Ask Michael?

File: src/MainProgram.cpp (hunk 20 above)
Comment (codeNinjaDev) [reply src/MainProgram.cpp:14]: We don't use it yet.

File: src/MainProgram.cpp (hunk 21 above)
Comment (codeNinjaDev) [reply src/MainProgram.cpp]: Michael?

File: src/Auto/Auto.cpp (hunk 5 above)
Comment (ajorgens) [reply src/Auto/Auto.cpp]: `stop(); //!! Check if this is necessary!!` starts the routine and timer?

File: src/RobotModel.cpp (hunk 24 above)
Comment (codeNinjaDev) [reply src/RobotModel.cpp]: Mike?

File: src/ShooterController.cpp (hunk 25 above)
Comment (codeNinjaDev) [reply src/ShooterController.cpp]: Mike?

**Pull request merged**